### PR TITLE
Allow to display/hide manufacturers and suppliers independently

### LINKED ISSUE #11332: Disable Supplier in front office
Hello,

I have the same request very regularly from E-merchants.

Most of them do not want suppliers with back office information to be visible in the front office.

Brands, of course, but not their suppliers.

Also, I think it should be disabled by default, and that an option could enable front-office visibility for those who want it.

### PR DESCRIPTION
**Pull request by matthieu-rolland**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Ability to display or hide suppliers and brands pages independantly
| Type?         | improvement
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | fixes #11332
| How to test?  | Everything is detailed in the issue.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/14665)

### CODE REVIEW

**Comment by matthieu-rolland** [comment]
Needed wording on the general configuration page

- Display suppliers
- Enable suppliers page on your front office even when its module is disabled.
- Display brands
- Enable brands page on your front office even when its module is disabled.

Needed wording on the brands / suppliers admin page

- The display of your brands is disabled/enabled on your store. Go to Shop Parameters > General to edit settings.
- The display of your suppliers is disabled/enabled on your store. Go to Shop Parameters > General to edit settings.

File: controllers/admin/AdminSuppliersController.php [hunk 1]
```diff
@@ -565,4 +567,23 @@ protected function afterUpdate($object)
 
         return true;
     }
+
+    /**
+     * supplierInfoMessage: builds a message related to the display suppliers configuration status
```
Comment (PierreRambaud) [inline controllers/admin/AdminSuppliersController.php]:
```suggestion
     * Builds a message related to the display suppliers configuration status
```

File: classes/controller/AdminController.php [hunk 2]
```diff
@@ -4882,4 +4882,19 @@ private function prepareNumberSpecifications(Context $context)
...
+     *
+     * @see admin-dev/themes/default/template/content.tpl
+     *
+     * @param string $message
+     *
+     * @return void
+     */
+    protected function setTopAlertInfoMessage($message)
+    {
+        $this->context->smarty->assign('alertInfo', $message);
```
Comment (PierreRambaud) [inline classes/controller/AdminController.php]: Isn't too overkill?
Should it be more generic?
```php
protected function addTopAlertMessage($message, $namespace = 'info') // Maybe use const here
{
    $messages = $this->context->smarty->getVariable('alertMessages')->value;
    if ($messages === null) {
        $messages = [];
    }

    if (!isset($messages[$namespace])) {
        $messages[$namespace] = [];
    }

    $messages[$namespace][] = $messages;

    $this->context->smarty->assign('alertMessages', $messages);
}
```

Comment (matthieu-rolland) [reply classes/controller/AdminController.php]: I don't think its overkill, the place where I was asked to put this info message text seems pretty generic to all legacy admin templates...  So it seems wrong to implement something specific coming only from AdminSuppliersController at this place.

But your suggestion is indeed more generic, I didn't think about all the kinds of info messages, I'll implement this !

Comment (matthieu-rolland) [reply classes/controller/AdminController.php]: done !

File: classes/controller/AdminController.php [hunk 3]
```diff
@@ -4882,4 +4894,30 @@ private function prepareNumberSpecifications(Context $context)
...
+    /**
+     * Using this method will display the given message as an "alert" message at the top of the current
+     * administration page.
+     *
+     * @see admin-dev/themes/default/template/content.tpl
+     *
+     * @param string $message The message to be displayed
+     * @param string $namespace Type of alert, use FRONT_ALERT_XXX constants
+     */
+    protected function addTopAlertMessage($message, $namespace = false)
```
Comment (PierreRambaud) [inline classes/controller/AdminController.php]:
```suggestion
    protected function addTopAlertMessage($message, $namespace = self::FRONT_ALERT_INFO)
```
and remove the `if (!$namespace) {` condition :)

File: controllers/admin/AdminPreferencesController.php [hunk 4]
```diff
@@ -194,6 +194,13 @@ public function __construct()
                     'cast' => 'intval',
                     'type' => 'bool',
                 ),
+                'PS_DISPLAY_MANUFACTURERS' => array(
+                    'title' => $this->trans('Display brands', array(), 'Admin.Shopparameters.Feature'),
+                    'desc' => $this->trans('Enable brands page on your front office even when its module are disabled.', array(), 'Admin.Shopparameters.Help'),
```
Comment (LouiseBonnard) [inline controllers/admin/AdminPreferencesController.php]:
```suggestion
                    'desc' => $this->trans('Enable brands page on your front office even when its module is disabled.', array(), 'Admin.Shopparameters.Help'),
```

File: controllers/admin/AdminPreferencesController.php [hunk 5]
```diff
@@ -194,6 +194,13 @@ public function __construct()
...
+                    'validation' => 'isBool',
+                    'cast' => 'intval',
+                    'type' => 'bool',
+                ),
                 'PS_DISPLAY_BEST_SELLERS' => array(
                     'title' => $this->trans('Display best sellers', array(), 'Admin.Shopparameters.Feature'),
                     'desc' => $this->trans('Enable best sellers page on your front office even when its respective module is disabled.', array(), 'Admin.Shopparameters.Help'),
```
Comment (LouiseBonnard) [inline controllers/admin/AdminPreferencesController.php]: Wouldn't it be the occasion to harmonize the help wording here? We could write _Enable best sellers page on your front office even when its module is disabled._ like the previous one.

File: controllers/admin/AdminSuppliersController.php [hunk 6]
```diff
@@ -565,4 +567,23 @@ protected function afterUpdate($object)
...
+     * @return string
+     */
+    protected function supplierInfoMessage()
+    {
+        $message = sprintf(
+            'The display of your suppliers is %s on your store. Go to Shop Parameters > General to edit settings.',
+            Configuration::get('PS_DISPLAY_SUPPLIERS') ? 'enabled' : 'disabled'
+        );
+
+        return $this->trans(
```
Comment (LouiseBonnard) [inline controllers/admin/AdminSuppliersController.php]: Is it okay if the wording does not follow the `$this->trans` here?

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php [hunk 7]
```diff
@@ -784,4 +785,14 @@ private function getAddressFormHandler()
...
     }
+
+    private function getAlertInfoMessage()
+    {
+        $message = sprintf(
+            'The display of your brands is %s on your store. Go to Shop Parameters > General to edit settings.',
+            $this->configuration->get('PS_DISPLAY_MANUFACTURERS') ? 'enabled' : 'disabled'
+        );
+
+        return $this->trans($message, 'Admin.Catalog.Notification');
```
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php]: Same question here. ;-)

File: controllers/admin/AdminPreferencesController.php (hunk 4 above)
Comment (LouiseBonnard) [reply controllers/admin/AdminPreferencesController.php]: We write this element twice? cf. https://github.com/PrestaShop/PrestaShop/pull/14665/files#diff-bc4c4faa966c237b7346ad189fdda6de

File: classes/controller/AdminController.php (hunk 3 above)
Comment (matthieu-rolland) [reply classes/controller/AdminController.php]: Good tip ! I was so persuaded I couldn't use a constant as a default parameter...

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php (hunk 7 above)
Comment (matthieu-rolland) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php]: Because the wording is dynamic, I build it first, inside the $message variable, then I use it inside $this->trans.
(is it what you meant ?)

File: controllers/admin/AdminSuppliersController.php (hunk 6 above)
Comment (matthieu-rolland) [reply controllers/admin/AdminSuppliersController.php]: Because the wording is dynamic, I build it first, inside the $message variable, then I use it inside $this->trans.
(is it what you meant ?)

Comment (LouiseBonnard) [reply controllers/admin/AdminSuppliersController.php]: Alright, perfect, I don't know much about code so I wanted to ask just to be sure. ;-)

**Comment by LouiseBonnard** [comment]
Two related suggestions from the community, about how to display manufacturers on a PrestaShop store:

http://feedback.prestashop.com/forums/387864/suggestions/36422122
http://feedback.prestashop.com/forums/387864/suggestions/36421930

@colinegin, @marionf, @samuel-pires

File: controllers/admin/AdminPreferencesController.php (hunk 5 above)
Comment (matthieu-rolland) [reply controllers/admin/AdminPreferencesController.php]: true, it's done !

File: controllers/admin/AdminPreferencesController.php [hunk 8]
```diff
@@ -188,15 +188,22 @@ public function __construct()
                     'class' => 'fixed-width-xxl',
                 ),
                 'PS_DISPLAY_SUPPLIERS' => array(
-                    'title' => $this->trans('Display brands and suppliers', array(), 'Admin.Shopparameters.Feature'),
-                    'desc' => $this->trans('Enable brands and suppliers pages on your front office even when their respective modules are disabled.', array(), 'Admin.Shopparameters.Help'),
+                    'title' => $this->trans('Display suppliers', array(), 'Admin.Shopparameters.Feature'),
+                    'desc' => $this->trans('Enable suppliers page on your front office even when its module are disabled.', array(), 'Admin.Shopparameters.Help'),
```
Comment (LouiseBonnard) [inline controllers/admin/AdminPreferencesController.php]:
```suggestion
                    'desc' => $this->trans('Enable suppliers page on your front office even when its module is disabled.', array(), 'Admin.Shopparameters.Help'),
```

File: controllers/admin/AdminPreferencesController.php (hunk 4 above)
Comment (matthieu-rolland) [reply controllers/admin/AdminPreferencesController.php]: yes, but it will not be displayed twice on the page, I just followed the existing logic for other fields in the same page and gave the translation key in this place as well.

File: classes/controller/AdminController.php (hunk 3 above)
Comment (jolelievre) [reply classes/controller/AdminController.php]: AS Pierre suggested you can use the constant directly as a default value

File: controllers/admin/AdminSuppliersController.php (hunk 6 above)
Comment (jolelievre) [reply controllers/admin/AdminSuppliersController.php]: Actually the `TranslationTool` won't be able to scan this message, because it can't `eval` the content of `$message` variable, si you'd better use an `if` condition and write two calls of `$this->trans` with hard coded strings

File: src/Adapter/Preferences/PreferencesConfiguration.php [hunk 9]
```diff
@@ -80,8 +81,10 @@ public function updateConfiguration(array $configuration)
             $this->configuration->set('PS_ROUND_TYPE', $configuration['price_round_type']);
             $this->configuration->set('PS_PRICE_DISPLAY_PRECISION', $configuration['price_display_precision']);
             $this->configuration->set('PS_DISPLAY_SUPPLIERS', $configuration['display_suppliers']);
+            $this->configuration->set('PS_DISPLAY_MANUFACTURERS', $configuration['display_manufacturers']);
             $this->configuration->set('PS_DISPLAY_BEST_SELLERS', $configuration['display_best_sellers']);
             $this->configuration->set('PS_MULTISHOP_FEATURE_ACTIVE', $configuration['multishop_feature_active']);
+            $this->configuration->set('PS_MULTISHOP_FEATURE_ACTIVE', $configuration['multishop_feature_active']);
```
Comment (jolelievre) [inline src/Adapter/Preferences/PreferencesConfiguration.php]: You have a duplicated line here

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php (hunk 7 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php]: Same thing here, yo need to use an `if` condition and two calls to `$this->trans` with hard coded strings.

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php [hunk 10]
```diff
@@ -100,6 +100,7 @@ public function indexAction(
             'help_link' => $this->generateSidebarLink($request->attributes->get('_legacy_controller')),
             'manufacturerGrid' => $this->presentGrid($manufacturerGrid),
             'manufacturerAddressGrid' => $this->presentGrid($manufacturerAddressGrid),
+            'alertInfoMessage' => $this->getAlertInfoMessage(),
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php]: I'm wondering if you couldn't use the flash messages for this, instead of adding a special variable and change the template.

File: controllers/admin/AdminSuppliersController.php [hunk 11]
```diff
@@ -67,6 +67,8 @@ public function __construct()
             'products' => array('title' => $this->trans('Number of products', array(), 'Admin.Catalog.Feature'), 'align' => 'right', 'filter_type' => 'int', 'tmpTableFilter' => true),
             'active' => array('title' => $this->trans('Enabled', array(), 'Admin.Global'), 'align' => 'center', 'active' => 'status', 'type' => 'bool', 'orderby' => false, 'class' => 'fixed-width-xs'),
         );
+
+        $this->addTopAlertMessage($this->supplierInfoMessage());
```
Comment (jolelievre) [inline controllers/admin/AdminSuppliersController.php]: This message is added on a legacy page, but maybe it's gonna be migrated during 177 development. We should update the migration issue so that this feature is not forgotten.

File: controllers/admin/AdminSuppliersController.php (hunk 6 above)
Comment (jolelievre) [reply controllers/admin/AdminSuppliersController.php]: I think it would be a good idea to integrate a link in the message, so that user can go directly on the appropriate page

Comment (jolelievre) [reply controllers/admin/AdminSuppliersController.php]: Also, with this the message will always be displayed, is there a way of closing this message?

**Review by jolelievre — CHANGES_REQUESTED** [review]
Nice job @matthieu-rolland 
I have a few technical feedback, and some about the feature because I remember discussing a lot about this issue ^^

File: controllers/admin/AdminSuppliersController.php (hunk 11 above)
Comment (jolelievre) [reply controllers/admin/AdminSuppliersController.php]: Oh actually the listing page has already been migrated but is hidden for now: https://github.com/PrestaShop/PrestaShop/issues/10545
You can access it by typing the url directly, so you should modify the migrated page rather than the legacy one

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php (hunk 10 above)
Comment (matthieu-rolland) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php]: I thought about it... here are the rasons why I did it this way, don't hesitate to tell me if you disagree:

1/ This is not really a "flash" message, because it is to be displayed permanently, so it shouldn't depend on session's flashbag. It is not supposed to be closed/removed either.

2/ There is currently no type of flash message for this "info" type of message, I used the already existing **infotip** helper in twig :
```twig
{{ ps.infotip(alertInfoMessage) }}
```
3/ I think the problem here is the name I gave to this variable and its method, "alertInfoMessage" is misleading, I should give it a more sensical name, like "permanentInfoMessage" and getPermanentInfoMessage.

What do you think?

File: controllers/admin/AdminSuppliersController.php (hunk 11 above)
Comment (matthieu-rolland) [reply controllers/admin/AdminSuppliersController.php]: yep, done !

File: controllers/admin/AdminSuppliersController.php (hunk 6 above)
Comment (matthieu-rolland) [reply controllers/admin/AdminSuppliersController.php]: Ok, I had no idea the translation messages were scanned without executing the code...
I made the fix.

Comment (matthieu-rolland) [reply controllers/admin/AdminSuppliersController.php]: About the link, yes I will add it ;)

Comment (jolelievre) [reply controllers/admin/AdminSuppliersController.php]: > Ok, I had no idea the translation messages were scanned without executing the code...

Indeed, of course I'm not talking about the `Translator` service, but `TranslatorTool` which we use to generate the XLF catalogs, it scan (almost) all the files of the projects (PHP, Twig, ...) looking code match `trans('my text')` or `'my text'|trans` and can then extract the translation keys. But of course it can not deduce the context that would generate a key if you integrated variables inside.

Comment (jolelievre) [reply controllers/admin/AdminSuppliersController.php]: About the link, you should use a key looking like this `To manage all your modules go to the [1]Installed module page[/1]` this prevents adding html code inside them

File: controllers/admin/AdminPreferencesController.php [hunk 12]
```diff
@@ -188,15 +188,22 @@ public function __construct()
                     'class' => 'fixed-width-xxl',
                 ),
                 'PS_DISPLAY_SUPPLIERS' => array(
-                    'title' => $this->trans('Display brands and suppliers', array(), 'Admin.Shopparameters.Feature'),
-                    'desc' => $this->trans('Enable brands and suppliers pages on your front office even when their respective modules are disabled.', array(), 'Admin.Shopparameters.Help'),
+                    'title' => $this->trans('Display suppliers', array(), 'Admin.Shopparameters.Feature'),
```
Comment (jolelievre) [inline controllers/admin/AdminPreferencesController.php]: Do you really need to change this legacy controller? I thought this page was already migrated to `/configure/shop/preferences/preferences`

File: controllers/admin/AdminSuppliersController.php (hunk 6 above)
Comment (jolelievre) [reply controllers/admin/AdminSuppliersController.php]: Here's an example https://github.com/PrestaShop/PrestaShop/blob/e762a65d41ff00588bcdccb1ee196bc6c835d4cc/src/PrestaShopBundle/Resources/views/Admin/Product/ProductPage/product.html.twig#L197

Although I think you can inject the parameters directly to `trans` instead of using `|replace`

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php (hunk 10 above)
Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php]: Alright that makes sense! Although I find it weird that you can never hide this message. We introduced in recent versions some help cards (or showcase) in some pages (which can be hidden thanks to cookies), maybe the same feature could have been used (although it's not up to you to decide)

Comment (jolelievre) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php]: About the name indeed it might be more understandable if you rename it, maybe `settingsInfoMessage` or `settingsTipMessage`

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/SupplierController.php [hunk 13]
```diff
@@ -500,4 +501,19 @@ private function getExceptionMessageByExceptionCode(Exception $exception)
 
         return $this->trans('Unexpected error occurred.', 'Admin.Notifications.Error');
     }
+
+    protected function getAlertInfoMessage()
```
Comment (jolelievre) [inline src/PrestaShopBundle/Controller/Admin/Sell/Catalog/SupplierController.php]: same renaming here as discussed on manufacturer page

File: controllers/admin/AdminSuppliersController.php (hunk 6 above)
Comment (matthieu-rolland) [reply controllers/admin/AdminSuppliersController.php]: Argh, the page didn't refresh with your last comment, I used another way (using %s in trans(), it seems to work, the translator actually does a sprintf)

File: controllers/admin/AdminPreferencesController.php (hunk 12 above)
Comment (matthieu-rolland) [reply controllers/admin/AdminPreferencesController.php]: I added it here fearing that not doing so might break something. If this controller is fully deprecated then yes, there's not need to modify it.

File: classes/controller/AdminController.php (hunk 3 above)
Comment (matthieu-rolland) [reply classes/controller/AdminController.php]: done !

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php (hunk 7 above)
Comment (matthieu-rolland) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php]: done !

File: src/Adapter/Preferences/PreferencesConfiguration.php (hunk 9 above)
Comment (matthieu-rolland) [reply src/Adapter/Preferences/PreferencesConfiguration.php]: fixed !

File: src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php (hunk 10 above)
Comment (matthieu-rolland) [reply src/PrestaShopBundle/Controller/Admin/Sell/Catalog/ManufacturerController.php]: done ! I gave it this name

File: controllers/admin/AdminPreferencesController.php (hunk 12 above)
Comment (jolelievre) [reply controllers/admin/AdminPreferencesController.php]: Now that it's done I think you can leave, but were you able to test it? Because if the routing is properly set with `_legacy_link` you shouldn't be able to access the legacy page any more

File: controllers/admin/AdminSuppliersController.php (hunk 6 above)
Comment (jolelievre) [reply controllers/admin/AdminSuppliersController.php]: yes it's better to use the trans parameter than replace, it's weird for the `[1]` we have a lot of them in our trads

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Manufacturer/index.html.twig [hunk 14]
```diff
@@ -25,22 +25,11 @@
...
-    icon: 'add_circle_outline',
-  },
-  add_manufacturer_address: {
-    href: path('admin_manufacturer_addresses_create'),
-    desc: 'Add new brand address'|trans({}, 'Admin.Catalog.Feature'),
-    icon: 'add_circle_outline',
-  },
-}
-%}
+
```
Comment (jolelievre) [inline src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Manufacturer/index.html.twig]: Why did you remove this part?

File: controllers/admin/AdminSuppliersController.php (hunk 6 above)
Comment (jolelievre) [reply controllers/admin/AdminSuppliersController.php]: You didn't manage the link in this legacy controller, this will result in two translation keys (one with the link placeholders, the other one without them)

File: src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Manufacturer/index.html.twig (hunk 14 above)
Comment (matthieu-rolland) [reply src/PrestaShopBundle/Resources/views/Admin/Sell/Catalog/Manufacturer/index.html.twig]: Damn, that's the result of a wrong manipulation on this file on my locale...  This is fixed now.

File: controllers/admin/AdminSuppliersController.php (hunk 6 above)
Comment (matthieu-rolland) [reply controllers/admin/AdminSuppliersController.php]: This legacy controller is no longer used so I just removed what I did here.

File: controllers/admin/AdminPreferencesController.php (hunk 12 above)
Comment (matthieu-rolland) [reply controllers/admin/AdminPreferencesController.php]: You're right, I couldn't test a page that is not linked. I just knew that the 'Display brands' translation had to be reworded, and I was not sure about the consequences if I left the old wordings untouched at this place.

**Comment by matthieu-rolland** [comment]
@jolelievre @matks  Addressed all review comments !

**Comment by PierreRambaud** [comment]
Sorry, you need to rebase your branch

**Comment by matthieu-rolland** [comment]
@PierreRambaud damn...yes 
I just did it

File: controllers/admin/AdminPreferencesController.php [hunk 15]
```diff
@@ -0,0 +1,253 @@
+<?php
```
Comment (jolelievre) [inline controllers/admin/AdminPreferencesController.php]: Seems like this file was removed from the branch, and you're re-adding it You can remove it from your branch as well

**Review by jolelievre — CHANGES_REQUESTED** [review]
One file to remove and we're good to go

**Review by jolelievre — DISMISSED** [review]

File: install-dev/upgrade/sql/1.7.7.0.sql [hunk 16]
```diff
@@ -2,3 +2,4 @@ SET SESSION sql_mode='';
 SET NAMES 'utf8';
 
 ALTER TABLE `PREFIX_search_word` MODIFY `word` VARCHAR(30) NOT NULL;
+INSERT INTO `PREFIX_configuration` (`name`, `value`, `date_add`, `date_upd`) VALUES('PS_DISPLAY_MANUFACTURERS', '1', NOW(), NOW());
```
Comment (PierreRambaud) [inline install-dev/upgrade/sql/1.7.7.0.sql]:
```suggestion
INSERT INTO `PREFIX_configuration` (`name`, `value`, `date_add`, `date_upd`) VALUES ('PS_DISPLAY_MANUFACTURERS', '1', NOW(), NOW());
```

**Comment by matthieu-rolland** [comment]
@PierreRambaud thanks for your input, I implemented your suggestion

File: src/PrestaShopBundle/Resources/views/Admin/macros.html.twig [hunk 17]
```diff
@@ -39,7 +39,7 @@
 {% macro infotip(text)%}
 <div class="alert alert-info" role="alert">
   <p class="alert-text">
-    {{ text }}
+    {{ text|raw }}
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: Any reason to use raw here?

Comment (matthieu-rolland) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: It is so that html is interpreted, otherwise when our translated message contains a link, that is to say a html tag (<a href ...) it will be displayed without being interpreted.

Comment (matthieu-rolland) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: @PierreRambaud would you prefer something like this?

`{{ text|striptags('<a>')|raw}}`

So that only `<a>` tags are accepted ?

Comment (matthieu-rolland) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: @PierreRambaud  up :)

Comment (matks) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: I think he would prefer no `raw` at all 😄

Do you know which usecases require to display such links ? It would be better to treat them as edge-cases with dedicated process. Something like:

```
{% if message.is_regular %}
    {{ text }}
{% elseif message.has.link %}
    {{ display_message_with_link(message) }}
{%  endif %}
```

Comment (PierreRambaud) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: If you change the behaviour by adding `raw`, you must be sure all variables using this macro are escaped.
Good idea from @matks :+1:

Comment (matthieu-rolland) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: The idea of `{{ text|striptags('<a>')|raw}}`was that striptags escapes everything except `<a>` tags, and the raw makes sure the non escaped `<a>` tag is interpretated. 

So all variables using this macro would be escaped (except their `<a>` tags).

But I like @matks 's solution too, I will do that.  👍

Comment (matks) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: 🤔I dont think scripttags is reliable
Here what we want to avoid are XSS vulnerabilities, and you can encode tags in a looooot of different ways. I dont think scripttags removes them all

Comment (matks) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: See https://www.owasp.org/index.php/XSS_Filter_Evasion_Cheat_Sheet

Comment (matthieu-rolland) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: @matks @PierreRambaud 
Ok I made a modification: instead of detecting if the link has a `<a>`tag or not, I made it possible to pass allowed tags as a parameter in the infotip macro function, if no allowed tag is specified then raw is not used at all, so that there are no breaking changes.

What do you think?

Comment (matthieu-rolland) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: damn I just saw your previous message @matks, so how do you suggest I display a variable having a `<a>` tag in it, without using raw ? 🤔

Comment (PierreRambaud) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: Striptags can remove unwanted characters

```
Hi, <:) we found it
```
will display
```
Hi,
```

Do we really need to add a link in the infotip?
Or do we want to add create a special infotip which is not escape? Maybe a `infotipraw`

Comment (matthieu-rolland) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: @PierreRambaud 

Well the link was not asked at the beginning, but suggested in this PR during the code review.
I do think it is useful for the end user though.

Do you mind if I add a raw parameter to the infotip function, whose default value is false, so that we don't duplicate the html inside the macro ?

Comment (matthieu-rolland) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: (that's what I did)

File: src/PrestaShopBundle/Resources/views/Admin/macros.html.twig [hunk 18]
```diff
@@ -36,10 +36,14 @@
   </span>
 {% endmacro %}
 
-{% macro infotip(text)%}
+{% macro infotip(text, raw = false)%}
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]:
```suggestion
{% macro infotip(text, use_raw = false)%}
```

To be sure there is zero conflict between `|raw` and the variable.

File: src/PrestaShopBundle/Resources/views/Admin/macros.html.twig (hunk 17 above)
Comment (PierreRambaud) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: If the link was suggested, I'm more in favor to remove it and keep the original behavior.
Otherwise if it's mandatory, adding a second parameter like you said if a good idea :).
At least it will be use only if necessary

File: src/PrestaShopBundle/Resources/views/Admin/macros.html.twig (hunk 18 above)
Comment (matthieu-rolland) [reply src/PrestaShopBundle/Resources/views/Admin/macros.html.twig]: @PierreRambaud done !

**Comment by matks** [comment]
@matthieu-rolland  it seems you wait for @PierreRambaud  validation of the changes you added following his code review, is that correct ?

**Comment by matthieu-rolland** [comment]
@matks 
Yep ! (up) 

@PierreRambaud 
I think you resolved my change, but didn't approve the PR as a whole ?

**Comment by PierreRambaud** [comment]
@matks @matthieu-rolland Still one issue unresolved. You checked it too fast :wink:

**Comment by matthieu-rolland** [comment]
> @matks @matthieu-rolland Still one issue unresolved. You checked it too fast 😉

@PierreRambaud  Indeed I didn't see the spacing fix suggestion, I made a commit out of it !

**Review by PierreRambaud — APPROVED** [review]
